planner/core/planbuilder.go

```diff
@@ -1510,41 +1512,82 @@ func (b *PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
 }
 
 func (b *PlanBuilder) buildDDL(node ast.DDLNode) (Plan, error) {
+	var authErr error
 	switch v := node.(type) {
 	case *ast.AlterTableStmt:
-		b.visitInfo = append(b.visitInfo, visitInfo{
-			privilege: mysql.AlterPriv,
-			db:        v.Table.Schema.L,
-			table:     v.Table.Name.L,
-			err:       nil,
-		})
+		if b.ctx.GetSessionVars().User != nil {
+			authErr = ErrTableaccessDenied.GenWithStackByArgs("ALTER", b.ctx.GetSessionVars().User.Hostname,
+				b.ctx.GetSessionVars().User.Username, v.Table.Name.L)
+		}
+		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.AlterPriv, v.Table.Schema.L,
+			v.Table.Name.L, "", authErr)
+		for _, spec := range v.Specs {
+			if spec.Tp == ast.AlterTableRenameTable {
+				if b.ctx.GetSessionVars().User != nil {
+					authErr = ErrTableaccessDenied.GenWithStackByArgs("DROP", b.ctx.GetSessionVars().User.Hostname,
+						b.ctx.GetSessionVars().User.Username, v.Table.Name.L)
+				}
+				b.visitInfo = appendVisitInfo(b.visitInfo, mysql.DropPriv, v.Table.Schema.L,
+					v.Table.Name.L, "", authErr)
+
+				if b.ctx.GetSessionVars().User != nil {
+					authErr = ErrTableaccessDenied.GenWithStackByArgs("CREATE", b.ctx.GetSessionVars().User.Hostname,
+						b.ctx.GetSessionVars().User.Username, spec.NewTable.Name.L)
+				}
+				b.visitInfo = appendVisitInfo(b.visitInfo, mysql.CreatePriv, spec.NewTable.Schema.L,
+					spec.NewTable.Name.L, "", authErr)
+
+				if b.ctx.GetSessionVars().User != nil {
+					authErr = ErrTableaccessDenied.GenWithStackByArgs("INSERT", b.ctx.GetSessionVars().User.Hostname,
+						b.ctx.GetSessionVars().User.Username, spec.NewTable.Name.L)
+				}
+				b.visitInfo = appendVisitInfo(b.visitInfo, mysql.InsertPriv, spec.NewTable.Schema.L,
+					spec.NewTable.Name.L, "", authErr)
+			} else if spec.Tp == ast.AlterTableDropPartition {
+				if b.ctx.GetSessionVars().User != nil {
+					authErr = ErrTableaccessDenied.GenWithStackByArgs("DROP", b.ctx.GetSessionVars().User.Hostname,
+						b.ctx.GetSessionVars().User.Username, v.Table.Name.L)
+				}
+				b.visitInfo = appendVisitInfo(b.visitInfo, mysql.DropPriv, v.Table.Schema.L,
+					v.Table.Name.L, "", authErr)
+			} else {
+				if b.ctx.GetSessionVars().User != nil {
+					authErr = ErrTableaccessDenied.GenWithStackByArgs("INSERT", b.ctx.GetSessionVars().User.Hostname,
```

Comment:
## In MySQL 5.7: 
```SQL
xiekeyi@ubuntu:~/gopath/src/github.com/pingcap/tidb$ mysql -u testuser -h localhost
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 26
Server version: 5.7.25-0ubuntu0.18.04.2 (Ubuntu)
Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.
Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.
Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
mysql> show grants for testuser;
+----------------------------------------------+
| Grants for testuser@%                        |
+----------------------------------------------+
| GRANT USAGE ON *.* TO 'testuser'@'%'         |
| GRANT ALTER ON `test`.`tt` TO 'testuser'@'%' |
+----------------------------------------------+
2 rows in set (0.00 sec)
mysql> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A
Database changed
mysql> alter table tt add column id3 int;
Query OK, 0 rows affected (0.10 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> alter table tt add index test_index(id3);
Query OK, 0 rows affected (0.03 sec)
Records: 0  Duplicates: 0  Warnings: 0

mysql> alter table tt;
Query OK, 0 rows affected (0.00 sec)

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 5.7.25-0ubuntu0.18.04.2 |
+-------------------------+
1 row in set (0.00 sec)
```
## In MySQL Doc:

> Enables use of the ALTER TABLE statement to change the structure of tables. ALTER TABLE also requires the CREATE and INSERT privileges. Renaming a table requires ALTER and DROP on the old table, CREATE, and INSERT on the new table.


Any advice for me?

Reply:
It seems like mysql doc is obsolete, but I think more strict privilege checking is better than less.

Reply:
Please update our doc to explain this.

Reply:
we'd better to follow the mysql behaviors, if users find they can alter table renaming to a new name in MySQL previously but can not perform this operation in TiDB after migration, that could be a problem

Reply:
```SQL
mysql> show grants for testuser;
+----------------------------------------------------+
| Grants for testuser@%                              |
+----------------------------------------------------+
| GRANT USAGE ON *.* TO 'testuser'@'%'               |
| GRANT SELECT ON `test`.* TO 'testuser'@'%'         |
| GRANT DROP, ALTER ON `test`.`tt` TO 'testuser'@'%' |
+----------------------------------------------------+
3 rows in set (0.00 sec)

mysql> show tables;
+----------------+
| Tables_in_test |
+----------------+
| jemp           |
| t1             |
| tt             |
| vt1            |
+----------------+
4 rows in set (0.00 sec)

mysql> ALTER TABLE t1;
ERROR 1142 (42000): ALTER command denied to user 'testuser'@'localhost' for table 't1'
mysql> ALTER TABLE t1 ADD INDEX index_name(id);
ERROR 1142 (42000): ALTER command denied to user 'testuser'@'localhost' for table 't1'
mysql> ALTER TABLE t1 rename to t2;
ERROR 1142 (42000): DROP, ALTER command denied to user 'testuser'@'localhost' for table 't1'
mysql> ALTER TABLE tt rename to tt2;
ERROR 1142 (42000): INSERT, CREATE command denied to user 'testuser'@'localhost' for table 'tt2'
mysql> 
```
Except `ALTER ... RENAME TO... ` and `ALTER ... DROP PATITION`,  `ALTER` is enough for others

-------

Our error messages seem not the same with MySQL when we have many privileges.
I will file an issue to track it.

Reply:
okay